[InstCombine] Fold fmod to frem if we know it does not set errno.

davemgreen opened the pull request:
fmod will be folded to frem in clang under -fno-math-errno and can be constant
folded in llvm if the operands are known. It can be relatively common to have
fp code that handles special values before doing some calculation:
```
if (isnan(f))
  return handlenan;
if (isinf(f))
  return handleinf;
..
fmod(f, 2.0)
```

This patch enables the folding of fmod to frem in instcombine if the first
parameter is not inf and the second is not zero. Other combinations do not set
errno.

nikic reviewed (commented):
This should be in SimplifyLibCalls.

nikic reviewed (commented):
This looks fine to me, but someone more familiar with FP should take a look at this...

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2278,8 +2278,8 @@ Value *LibCallSimplifier::replacePowWithSqrt(CallInst *Pow, IRBuilderBase &B) {
   if (!Pow->doesNotAccessMemory() && !Pow->hasNoInfs() &&
-      !isKnownNeverInfinity(Base, 0,
-                            SimplifyQuery(DL, TLI, /*DT=*/nullptr, AC, Pow)))
+      !isKnownNeverInfinity(
+          Base, 0, SimplifyQuery(DL, TLI, DT, AC, Pow, true, true, DC)))
```
Passing in the new analyses are a separate patch

davemgreen replied:
I'll rebase over #108446 once it goes in, probably tomorrow morning.

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2796,6 +2796,25 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
+  if (Known0.isKnownNeverInfinity()) {
+    KnownFPClass Known1 = computeKnownFPClass(CI->getOperand(1), fcZero,
+                                              /*Depth=*/0, SQ);
+    if (Known1.isKnownNeverZero()) {
```
This is wrong if DAZ is enabled, this needs to account of the denormal mode. Can you add a negative test with "denormal-fp-math"="preserve-sign,preserve-sign" and/or ="dynamic,dynamic"? This should be checking isKnownNeverLogicalZero

davemgreen replied:
Thanks - my testing (https://gist.github.com/davemgreen/f6b084618c24db2b0996cc618cd8ca05) was on a normal linux system. There is a test in the end of ad3ad15229cc65874eae173137e47b5c817d77ea, and I've changed it to isKnownNeverLogicalZero.

davemgreen committed:
[InstCombine] Fold fmod to frem if we know it does not set errno.

fmod will be folded to frem in clang under -fno-math-errno and can be constant
folded in llvm if the operands are known. It can be relatively common to have
fp code that handles special values before doing some calculation:
```
if (isnan(f))
  return handlenan;
if (isinf(f))
  return handleinf;
..
fmod(f, 2.0)
```

This patch enables the folding of fmod to frem in InstCombine (via
LibCallSimplifier) if the first parameter is not inf and the second is not
zero. Other combinations do not set errno.

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2796,6 +2796,27 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
+  KnownFPClass Known0 = computeKnownFPClass(CI->getOperand(0), fcInf,
+                                            /*Depth=*/0, SQ);
+  if (Known0.isKnownNeverInfinity()) {
+    KnownFPClass Known1 = computeKnownFPClass(CI->getOperand(1), fcZero,
```
fcZero | fcSubnormal

arsenm commented on llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp:
```diff
@@ -2796,6 +2796,27 @@ Value *LibCallSimplifier::optimizeSqrt(CallInst *CI, IRBuilderBase &B) {
+
+  // fmod(x,y) can set errno if y == 0 or x == +/-inf. Otherwise we can change
+  // it to a frem instruction.
+  KnownFPClass Known0 = computeKnownFPClass(CI->getOperand(0), fcInf,
```
You can also just do this if the instruction has the nnan flag

arsenm reviewed (approved):
LGTM, but this is ignoring the flags directly on the instruction. If I read this correctly, you can just do this for any nnan call

davemgreen committed:
Change to test and set NoNan

davemgreen commented:
I've had a go at making it work of nnan instructions too. Let me know if this still looks OK.
